### ISSUE TITLE
The empty object is a valid test provider

### ISSUE THREAD

**Issue by jrieken** [microsoft/vscode#117403]
Testing #117307

tl;dr: This should not compile: `vscode.test.registerTestProvider({})`. 

An API should provide guidance, e.g it should force you to implement something and provide rigid rules on how to get to a result. It seems the `TestProvider`-type is "Jack of all trades" and we should define its exact role. Is it a provider? Is is a runner? Is it workspace test provider or just a document test provider etc. 

So, one thing that I wonder is why the test provider doesn't fit into the usual provider pattern, e.g the usual triple of provideX, resolveX, and onDidChangeX

```ts
export interface TestItemProvider {

  // signal that something changes, per document/globally
  onDidChangeTestItems?: Event<TBD>

  // return test data from which we create test domain objects
  provideTestItems(folder: WorkspaceFolder | undefined, document: TextDocument | undefined, token?: CancellationToken): ProviderResult<TestItem[]>

  // optionally for heavy lifting before running(?), expanding(?)
  resolveTestItem?(item: TestItem): ProviderResult<TestItem>
}
```

Then, it should be considered to decouple "test running" from "test providing". Despite high chances of being implemented by the same extension it might add clarity if discovery and running are "described" with separate types.

**Comment by connor4312** [comment]
I think it makes sense for the workspace test method to be required.

You might have seen the `@todo` I have to rename the testprovider to be not a provider. I think the didChange/provide methods you propose are feasible with the way things are now, but I would still want some additional data from the "provide", at minimum a promise or event to indicate the state of discovery. (The alternative in this design is to return a Promise from provideTestItems, but the downside here is that tests would not be visible until they're all discovered.)

We would also have a separate invalidated event on the provider.

--

The notion behind the "test provider" being the one running tests is that we can provide the same object instances that was given by the provider to the runner. These objects could, and in the two providers so far do, have extra methods and metadata used for running them. If there's a separate runner, there's no guarantee that it's running in the same extension host, so we would provide only the serialized version of tests to the runners.

It also either requires us to add API that lets extensions figure out which provider provided which tests, or requires extensions to figure out if the given instance is runnable by this test runner. We would also need to deal with the possibility of a single test in one run being executed by multiple runners in the same test run, which is not something we deal with today.

**Comment by connor4312** [comment]
However, I will make the workspace test providers and runTests methods required (even if runTests is later moved elsewhere...)

**Comment by connor4312** [comment]
- We now require `runTests` and `createWorkspaceTestHierarchy`
- In https://github.com/microsoft/vscode/commit/17f63a649ec3831f7d9144746dd947f42600c9d2 I've update the `create` methods to more proper `provide` methods.
- "test running" is being discussed in https://github.com/microsoft/vscode/issues/117376
